nsmdp/nsmdp.go

```diff
@@ -20,89 +20,123 @@
 package nsmdp
 
 import (
-	"context"
-	"crypto/rand"
 	"fmt"
+	"net"
 	"os"
+	"path"
+	"strconv"
+	"time"
 
-	"github.com/ligato/networkservicemesh/deviceplugin"
+	"github.com/ligato/cn-infra/logging"
+
+	"golang.org/x/net/context"
+	"golang.org/x/sys/unix"
+	"google.golang.org/grpc"
 	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1beta1"
 )
 
 const (
-	SocketBaseDir = "/var/lib/networkservicemesh/"
+	socketBaseDir   = "/var/lib/networkservicemesh/"
+	resourceName    = "nsm.ligato.io/socket"
+	serverSock      = "nsm.ligato.io.sock"
+	initDeviceCount = 10
 )
 
-type NSMDevicePlugin struct {
-	*deviceplugin.DevicePlugin
-	devs       map[string]*NSMDevice
-	updatedevs chan *NSMDevice
-	stop       chan interface{}
+type nsmSocket struct {
+	device      *pluginapi.Device
+	socketPath  string
+	stopChannel chan bool
+	allocated   bool
 }
 
-type NSMDevice struct {
-	*pluginapi.Device
-	token string
+type nsmClientEndpoints struct {
+	nsmSockets map[string]nsmSocket
+	logger     logging.PluginLogger
 }
 
-const (
-	resourceName    = "nsm.ligato.io/socket"
-	serverSock      = pluginapi.DevicePluginPath + "nsm.ligato.io.sock"
-	initDeviceCount = 10
-)
+// Define functions needed to meet the Kubernetes DevicePlugin API
+func (n *nsmClientEndpoints) GetDevicePluginOptions(context.Context, *pluginapi.Empty) (*pluginapi.DevicePluginOptions, error) {
+	n.logger.Infof(" GetDevicePluginOptions was called.")
+	return &pluginapi.DevicePluginOptions{}, nil
+}
 
-func NewNSMDevicePlugin() *NSMDevicePlugin {
-	n := &NSMDevicePlugin{
-		DevicePlugin: deviceplugin.NewDevicePlugin(serverSock, resourceName),
-		devs:         make(map[string]*NSMDevice),
-		stop:         make(chan interface{}),
-		updatedevs:   make(chan *NSMDevice, 10),
-	}
-	for i := uint(0); i < initDeviceCount; i++ {
-		dev := &NSMDevice{
-			Device: &pluginapi.Device{
-				ID:     "NSM_" + string(i),
-				Health: pluginapi.Healthy,
-			},
-			token: generateToken(),
-		}
-		n.updatedevs <- dev
+func startClientServer(client *nsmSocket, logger logging.PluginLogger) {
+	logger.Infof(" startClientServer was called with socket: %s", client.socketPath)
+	listenEndpoint := client.socketPath
+	fi, err := os.Stat(listenEndpoint)
+	if err == nil && (fi.Mode()&os.ModeSocket) != 0 {
+		os.Remove(listenEndpoint)
+	}
+	if err != nil && !os.IsNotExist(err) {
+		logger.Infof(" failure stat of socket file %s with error: %+v", client.socketPath, err)
+		client.allocated = false
+		return
 	}
-	return n
-}
 
-func generateToken() string {
-	b := make([]byte, 16)
-	rand.Read(b)
-	return fmt.Sprintf("%x", b)
-}
+	unix.Umask(0077)
+	sock, err := net.Listen("unix", listenEndpoint)
+	if err != nil {
+		logger.Infof(" failure to listen on socket %s with error: %+v", client.socketPath, err)
```

Comment:
Nit: Space before failure should be removed. I see this in a few logs below. And I think we should be consistent and capitalize the first word in a log message as well.